trustfall_core/src/interpreter/execution.rs

```diff
@@ -318,9 +362,28 @@ fn collect_fold_elements<'query, Vertex: Clone + Debug + 'query>(
 
         Some(fold_elements)
     } else {
-        // We weren't able to find any early-termination condition for materializing the fold,
-        // so materialize the whole thing and return it.
-        Some(iterator.collect())
+        let collected = match min_fold_count_limit {
+            // If we have a min_fold_count_limit then it is safe to .take(min_fold_count_limit)
+            // because .take will take at max min_fold_count_limit elements, and if the iterator has
+            // less it will just take all of them.
```

Comment:
```suggestion
            // If we have a min_fold_count_limit then it is safe to .take(min_fold_count_limit)
            // because it will take at most min_fold_count_limit elements, and if the iterator has
            // fewer items then it will just take all of them.
```